README.md
VariConf - Load Configurations from Various Types of Files
==========================================================

VariConf provides a wrapper around [OmegaConf](https://omegaconf.readthedocs.io) for
loading configuration from various types of files.

Supported file types are JSON, YAML and TOML.  Support for more file types can easily be
added by registering a custom loader function.

Thanks to the power of OmegaConf, you can provide a configuration schema which the
defines expected parameters, default values and optionally expected types.


Design Goals
------------

This package is developed with the following goals in mind:

- Load configuration from files with expected parameters and default values provided in
  an easy way.
- Provide config as a simple Namespace-like object (with option to convert to
  dictionary).
- Do not commit to a specific file format.  All formats that can easily be loaded into a
  dictionary should be supported (json, toml, yaml, ...).
- Optionally check types.
- Optionally check for unknown parameters in the file (to raise an error).
- Keep it simple.  Prefer fewer features over too complicated API.


Installation
------------

Basic (does not include dependencies for YAML and TOML):
```
pip install variconf
```

With optional dependencies:
```bash
# with TOML support
pip install "variconf[toml]"

# with YAML support
pip install "variconf[yaml]"

# to include everything:
pip install "variconf[all]"
```


Usage
-----

The package provides a class `WConf` for loading and merging configurations from
different sources. When creating an instance of it, a configuration "schema" needs to be
given, i.e. a structure (dictionary or dataclass) that defines what sections and
parameters the configuration has and that provides default values.

A number of "load"-methods is provided to load configurations from different sources
(e.g. JSON files, YAML files, dictionaries from other sources, ...).  When called,
the corresponding parameters are merged into the existing configuration, overwriting
existing values.  This means that an input does not need to provide all parameters,
in this case the default values are kept.  Further, if calling multiple
"load"-methods after another, the later calls will overwrite values set by previous
ones.

All "load"-methods return ``self``, so they can be chained:

```python
import variconf

schema = {"sec1": {"foo": 42, "bar": 13}, "sec2": {"bla": ""}}
wconf = variconf.WConf(schema)
config = (
    wconf.load_file("~/global_config.toml")
    .load_file("./local_config.yml", fail_if_not_found=False)
    .load_dotlist(sys.argv[1:])  # easily allow overwriting parameters via
                                 # command-line arguments
    .get()  # return the final config object
)
```


### Allow Unknown Parameters

By default an error is raised if the loaded configuration contains parameters that are
not declared in the schema.  If you want to allow these unknown parameters, initialise
`WConf` with `strict=False`:

```python
wconf = variconf.WConf(schema, strict=False)
```

This will result in the unknown parameters being merged into the config object.

With this you can even omit the schema altogether by simply passing an empty dictionary:
```python
wconf = variconf.WConf({}, strict=False)
```


### Search File

Assuming an application where the config file can be located in one of several places
(e.g. `~`, `~/.config` or `/etc/myapp`).  This situation is supported by the optional
`search_paths` argument of `load_file()`:

```python
wconf.load_file(
    "config.yml",
    search_paths=[os.expanduser("~"), os.expanduser("~/.config"), "/etc/myapp"],
    fail_if_not_found=False,
)
```
This will search for a file "config.yml" in the listed directories (in the given order)
and use the first match.
By setting `fail_if_not_found=False`, we specify that it's okay if the file is not found
in any of these directories.  In this case, we simply keep the default values of all
parameters.


### Using XDG Base Directory Specification

If your application follows the [XDG Base Directory
Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html)
you can use ``load_xdg_config()`` (currently not supported on Windows!):

```python
wconf.load_xdg_config("myapp/config.toml")
```
Will search for the file in the directories specified in the environment variables
`XDG_CONFIG_HOME` and `XDG_CONFIG_DIRS` (defaulting to `~/.config`).

Like for `load_file()` there is an argument `fail_if_not_found` but here it defaults to
False as providing a config in `XDG_CONFIG_HOME` is typically optional.


### Supported File Types

Supported file types are JSON, YAML and TOML.  Support for custom file types can be
added by providing a loader function.  Example for adding XML support:

```python
import xml.etree.ElementTree as ET

def xml_loader(fp: typing.IO) -> dict:
    xml_str = fp.read()
    xml_tree = ET.fromstring(xml_str)
    # do some magic to convert XML tree to dictionary
    xml_dict = tree_to_dict(xml_tree)
    return xml_dict

wconf.add_file_loader("xml", [".xml"], xml_loader)

# now, XML files can be read by WConf.load and WConf.load_file
wconf.load_file("config.xml")
```


### Type Checking

OmegaConf supports type-checking by providing a schema as dataclass with type hints:

```python
@dataclasses.dataclass
class ConfigSchema:
    foo: int = 42
    bar: str = 13

wconf = variconf.WConf(ConfigSchema)

# raises ValidationError: Value 'hi' of type 'str' could not be converted to Integer
wconf.load_dict({"foo": "hi"})
```

### Required Values

Required parameters without default value are supported through OmegaConf's concept of
missing values.

When using a dictionary schema:

```python
schema = {
    "optional_param": "default value",
    "required_param": "???",
}
```

When using a dataclass schema:

```python
@dataclasses.dataclass
class Schema:
    required_param1: float  # not providing a default makes it required
    optional_param: str = "default value"
    required_param2: int = omegaconf.MISSING  # alternative for required parameters
```

If there is a required parameter for which no value has been provided by any of the
`load*`-methods, calling `get()` will raise an error.

You can avoid that error by using `get(allow_missing=True)`.  However, the error is
still raised when trying to access the actual value of the missing parameter.


### Variable Interpolation

OmegaConf has a feature called [variable interpolation](https://omegaconf.readthedocs.io/en/latest/usage.html#variable-interpolation)
that allows to refer to other fields within the config file:

```yaml
server:
  host: localhost
  port: 80

client:
  url: http://${server.host}:${server.port}/
  server_port: ${server.port}
  # relative interpolation
  description: Client of ${.url}
```
See the documentation of OmegaConf for more information.


Copyright and License
---------------------

VariConf is written and maintained by Felix Kloss at the Max Planck Institute for
Intelligent Systems, Tübingen.

Copyright (c) 2022, Max Planck Gesellschaft. All rights reserved.

License: [BSD 3-clause](LICENSE)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]


## [1.0.1] - 2023-02-10
### Fixed
- Install `py.typed` so type annotations can be used in other packages.


## [1.0.0] - 2022-11-02

First proper release.


[Unreleased]: https://github.com/MPI-IS/variconf/compare/1.0.0...HEAD
[1.0.0]: https://github.com/MPI-IS/variconf/releases/tag/v1.0.0


variconf/wconf.py
# Copyright (c) 2022 Max Planck Gesellschaft
# SPDX-License-Identifier: BSD-3-Clause
from __future__ import annotations

import typing
import pathlib
import platform
import os

import omegaconf as oc

from . import errors


_PathLike = typing.Union[os.PathLike, str]

LoaderFunction = typing.Union[
    typing.Callable[[typing.IO], dict],
    typing.Callable[[typing.TextIO], dict],
    typing.Callable[[typing.BinaryIO], dict],
]


def find_file(
    filename: pathlib.Path, search_paths: typing.Sequence[_PathLike]
) -> pathlib.Path:
    """Search for a file in a list of directories.

    Args:
        filename: Name of the file (can also be a relative path).
        search_paths:  List of directories in which the file is searched.

    Returns:
        The path to the first matching file that is found.

    Raises:
        FileNotFoundError: If the file is not found in any of the directories in
            search_paths.
    """
    for directory in map(pathlib.Path, search_paths):
        _file = directory / filename
        if _file.is_file():
            return _file

    raise FileNotFoundError(f"No file {filename} found in {search_paths}")


class WConf:
    """Load configuration from files.

    A thin wrapper around OmegaConf to easily load configurations from various types of
    files.

    When creating an instance of ``WConf``, a configuration "schema" needs to be given,
    i.e. a structure (e.g. dictionary) that defines what sections and parameters the
    configuration has and that provides default values.

    A number of "load"-methods is provided to load configurations from different sources
    (e.g. JSON files, YAML files, dictionaries from other sources, ...).  When called,
    the corresponding parameters are merged into the existing configuration, overwriting
    existing values.  This means that an input does not need to provide all parameters,
    in this case the default values are kept.  Further, if calling multiple
    "load"-methods after another, the later calls will overwrite values set by previous
    ones.

    All "load"-methods return ``self``, so they can be chained:

    .. code-block:: python

        schema = {"sec1": {"foo": 42, "bar": 13}, "sec2": {"bla": ""}}
        wconf = WConf(schema)
        config = (
            wconf.load_file("~/global_config.toml")
            .load_file("./local_config.yml")
            .load_dotlist(sys.argv[1:])
            .get()  # return the final config object
        )

    """

    def __init__(self, schema, strict: bool = True) -> None:
        """
        Args:
            schema: Configuration structure including all parameters with their default
                vaules.  This can be a dictionary, a dataclass (in which case type
                checking is enabled) or any other object that is supported by
                ``OmegaConf.create``.
            strict: If true, loading a configuration with parameters that are not listed
                in schema results in an error.  If false, they will be merged into the
                configuration.
        """
        self.cfg = oc.OmegaConf.create(schema)

        if strict:
            oc.OmegaConf.set_struct(self.cfg, True)

        self._loaders: typing.Dict[str, typing.Tuple[LoaderFunction, bool]] = {
            "json": (self._load_json, False),
            "toml": (self._load_toml, True),
            "yaml": (self._load_yaml, False),
        }
        # map extension (including dot) to format type
        self._file_extensions = {
            ".json": "json",
            ".toml": "toml",
            ".yml": "yaml",
            ".yaml": "yaml",
        }

    def _merge(self, cfg) -> None:
        self.cfg = oc.OmegaConf.merge(self.cfg, cfg)

    def _load_json(self, fp: typing.IO) -> dict:
        import json

        return json.load(fp)

    def _load_yaml(self, fp: typing.IO) -> dict:
        import yaml

        return yaml.safe_load(fp)

    def _load_toml(self, fp: typing.BinaryIO) -> dict:
        import tomli

        return tomli.load(fp)

    def get_supported_formats(self) -> typing.List[str]:
        """Get list of supported file formats."""
        return list(self._loaders.keys())

    def get(self, allow_missing=False) -> oc.OmegaConf:
        """Get the configuration object.

        Args:
            allow_missing: If enabled, the config can still contain required parameters
                which have not been set (accessing them will result in an error).  If
                disabled (default), get() will check the configuration for missing
                required values and raise an error instead of returning the config.

        Raises:
            omegaconfig.errors.MissingMandatoryValue: If ``allow_missing=False`` and the
                config has one or more required values (marked as missing in the
                schema), which have not been filled yet.
        """
        if not allow_missing:
            # the easiest way to check the whole config is to convert to a container
            # with throw_on_missing=True (even though, we're not interested in the
            # resulting container...).
            oc.OmegaConf.to_container(self.cfg, throw_on_missing=True)

        return self.cfg

    def load(self, fp, format: str) -> WConf:
        """Load configuration from the given stream.

        See :meth:`get_supported_formats` for a list of supported file formats.  Custom
        loaders for other formats can be added with :meth:`add_file_loader`.

        Args:
            fp: Input stream containing a document in the specified format.
            format: Name of the format (e.g. "yaml").

        Returns:
            ``self``, so methods can be chained when loading from multiple sources.
        """
        try:
            loader, _ = self._loaders[format]
        except KeyError:
            raise errors.UnknownFormatError(format)

        cfg = loader(fp)
        self._merge(cfg)

        return self

    def load_file(
        self,
        file: _PathLike,
        fail_if_not_found: bool = True,
        search_paths: typing.Optional[typing.Sequence[_PathLike]] = None,
    ) -> WConf:
        """Load configuration from the specified file.

        The format of the file is derived from the filename extension.
        See :meth:`get_supported_formats` for a list of supported file formats.  Custom
        loaders for other formats can be added with :meth:`add_file_loader`.

        Args:
            file:  A file in one of the supported formats.
            fail_if_not_found:  If true, raise an error if the file is not found.
                Otherwise simply return without loading anything (i.e. keep the current
                values).
            search_paths:  List of directories.  If set, search these directories for a
                file with the name specified in ``file`` and loads the first file that
                is found.

        Returns:
            ``self``, so methods can be chained when loading from multiple sources.
        """
        file = pathlib.Path(file)

        # If search_path is set, search for the file in these paths.  Otherwise directly
        # use `file`.
        try:
            if search_paths:
                file = find_file(file, search_paths)
            else:
                if not file.is_file():
                    raise FileNotFoundError(file)
        except Exception as e:
            if fail_if_not_found:
                raise e
            else:
                return self

        try:
            fmt = self._file_extensions[file.suffix]
        except KeyError:
            raise errors.UnknownExtensionError(file.suffix)

        _, binary = self._loaders[fmt]
        if binary:
            mode = "rb"
        else:
            mode = "rt"

        with open(file, mode) as f:
            self.load(f, fmt)

        return self

    def add_file_loader(
        self,
        name: str,
        file_extensions: typing.Sequence[str],
        loader: LoaderFunction,
        binary: bool = False,
    ) -> None:
        """Add a custom file loader.

        Add a custom loader to add support for other file formats.  It is also possible
        to overwrite the default loaders for existing formats.

        Args:
            name:  Name of the format.
            file_extensions:  List of file extensions by which files of this format can
                be detected.  The extensions must include the leading dot!
            loader:  Function that takes an file-like object as input and returns a
                dictionary.
            binary:  Set to True if the loader function expects a binary stream.
        """
        self._loaders[name] = (loader, binary)
        self._file_extensions.update({ext: name for ext in file_extensions})

    def load_object(self, config) -> WConf:
        """Load configuration from a dictionary, dataclass or OmegaConf instance.

        Args:
            config: Object containing the configuration.  See OmegaConf.merge for
                supported types.

        Returns:
            ``self``, so methods can be chained when loading from multiple sources.
        """
        self._merge(config)
        return self

    def load_dict(self, config: typing.Mapping) -> WConf:
        """Alias for :meth:`load_object`."""
        return self.load_object(config)

    def load_dotlist(self, dotlist: typing.List[str]) -> WConf:
        """Load configuration from a "dotlist"

        A dotlist looks like this: ``["foo.bar=42", "baz=13", ...]``
        This can, for example, be used to allow overwriting parameters from the file via
        a command-line argument.

        Args:
            dotlist: List of parameters in dot-notation.

        Returns:
            ``self``, so methods can be chained when loading from multiple sources.
        """
        cfg = oc.OmegaConf.from_dotlist(dotlist)
        self._merge(cfg)
        return self

    @staticmethod
    def _get_xdg_config_paths() -> typing.List[pathlib.Path]:
        if platform.system() == "Windows":
            raise NotImplementedError("XDG support is not implemented for Windows.")

        # get config_home
        config_home_default = pathlib.Path(os.environ["HOME"], ".config")

        config_home = pathlib.Path(os.environ.get("XDG_CONFIG_HOME", ""))
        if not config_home.is_absolute():
            config_home = config_home_default

        # get config dirs
        config_dirs_str = os.environ.get("XDG_CONFIG_DIRS", "")
        _config_dirs = config_dirs_str.split(":")
        config_dirs = list(
            filter(lambda p: p.is_absolute(), map(pathlib.Path, _config_dirs))
        )
        if not config_dirs:
            config_dirs = [pathlib.Path("/etc/xdg")]

        return [config_home] + config_dirs

    def load_xdg_config(self, filename: _PathLike, fail_if_not_found=False) -> WConf:
        """Load file from XDG config directory.

        Searches for the specified file in the directories given in the
        ``XDG_CONFIG_HOME`` and ``XDG_CONFIG_DIRS`` environment variables.  If they are
        not set, they default to ``${HOME}/.config`` and ``/etc/xdg``.

        For more information on the XDG base directory specification see
        https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

        Args:
            filename: Name of the file, relative to the config directory.
            fail_if_not_found:  If true, raise an error if the file is not found.
                Otherwise simply return without loading anything (i.e. keep the current
                values).
        """
        paths = self._get_xdg_config_paths()
        return self.load_file(
            filename, fail_if_not_found=fail_if_not_found, search_paths=paths
        )
